File: ingestion/common/rate_limit.py
"""Simple minimum-interval rate limiter.

The first call never waits. Subsequent calls wait just enough to keep calls
at least ``min_interval_seconds`` apart. ``clock`` / ``sleep`` are injectable
for fast, deterministic tests.
"""

from __future__ import annotations

import time
from typing import Callable, Optional


class RateLimiter:
    def __init__(
        self,
        min_interval_seconds: float = 0.0,
        *,
        clock: Callable[[], float] = time.monotonic,
        sleep: Callable[[float], None] = time.sleep,
    ) -> None:
        self._min = max(0.0, float(min_interval_seconds))
        self._clock = clock
        self._sleep = sleep
        self._last: Optional[float] = None

    @property
    def min_interval(self) -> float:
        return self._min

    def acquire(self) -> float:
        """Block (via injected sleep) until enough time elapsed. Returns wait."""
        if self._min <= 0:
            self._last = self._clock()
            return 0.0
        now = self._clock()
        waited = 0.0
        if self._last is not None:
            elapsed = now - self._last
            if elapsed < self._min:
                wait = self._min - elapsed
                self._sleep(wait)
                waited = wait
        self._last = self._clock()
        return waited

File: platform/workers/ingestion-premium/src/rate_limit.ts
/**
 * Phase 3.5 (P0-4) — Promise-based minimum-interval rate limiter.
 *
 * The limiter serializes state mutation through a Promise chain so that
 * concurrent ``acquire()`` callers each receive a slot at least
 * ``min_interval_ms`` apart — even though JavaScript is single-threaded, the
 * event loop interleaves ``fetch`` resolutions, so a shared reservation
 * variable must be guarded.
 *
 * Behavior:
 *   * First call never waits (reserves the next slot at now + interval).
 *   * Subsequent calls either take the reserved slot immediately or sleep
 *     until it is their turn, then advance the reservation by one interval.
 *   * ``min_interval_ms <= 0`` disables the limiter (``acquire`` is a no-op).
 *   * On 429: ``notify429()`` applies a short cooldown and temporarily widens
 *     the interval; subsequent successes via ``notifyOk()`` restore the base
 *     floor quickly so we return to near-ceiling throughput.
 *
 * Premium budget ~500 req/min → 120 ms base floor (exactly 500/min).
 */
export class RateLimiter {
  private readonly baseIntervalMs: number;
  private currentIntervalMs: number;
  private nextAllowedMs = 0;
  private tail: Promise<void> = Promise.resolve();
  private consecutiveOk = 0;
  private rateLimitHits = 0;

  constructor(minIntervalMs: number) {
    this.baseIntervalMs = Math.max(0, Math.floor(minIntervalMs));
    this.currentIntervalMs = this.baseIntervalMs;
  }

  get minIntervalMsValue(): number {
    return this.currentIntervalMs;
  }

  get baseIntervalMsValue(): number {
    return this.baseIntervalMs;
  }

  get rateLimitHitCount(): number {
    return this.rateLimitHits;
  }

  /**
   * Short 429 recovery: push next slot ~1.2s out and briefly run at 2× interval.
   * Does not park at a low long-term rate — ``notifyOk`` restores base quickly.
   */
  notify429(cooldownMs = 1_200): void {
    this.rateLimitHits += 1;
    this.consecutiveOk = 0;
    if (this.baseIntervalMs > 0) {
      // 2× base (e.g. 240 ms) — still near the Premium class, not a deep park.
      this.currentIntervalMs = Math.max(this.baseIntervalMs * 2, this.baseIntervalMs);
    }
    const now = Date.now();
    this.nextAllowedMs = Math.max(this.nextAllowedMs, now + Math.max(0, cooldownMs));
  }

  /** After a successful upstream response, decay back toward the base floor. */
  notifyOk(): void {
    this.consecutiveOk += 1;
    if (this.consecutiveOk >= 2 && this.currentIntervalMs > this.baseIntervalMs) {
      this.currentIntervalMs = this.baseIntervalMs;
    }
  }

  async acquire(): Promise<void> {
    if (this.baseIntervalMs <= 0) return;
    // Chain a reservation: each caller waits for the previous to finish
    // (which has already advanced nextAllowedMs past its own slot).
    const slot = this.tail.then(async () => {
      const now = Date.now();
      const wait = Math.max(0, this.nextAllowedMs - now);
      if (wait > 0) {
        await new Promise<void>((resolve) => setTimeout(resolve, wait));
      }
      const interval = this.currentIntervalMs > 0 ? this.currentIntervalMs : this.baseIntervalMs;
      this.nextAllowedMs = Date.now() + interval;
    });
    // Keep the chain ordered; do not let a thrown rejection break later slots.
    this.tail = slot.then(
      () => undefined,
      () => undefined,
    );
    await slot;
  }
}
